File: tools/motorCAD/scaling/SLScaleMachine.m


function ScaledMachineData = SLScaleMachine(scalingFactorStruct,MotorCADGeo,BuildingData)
    % MotorCADGeo =refLABBuildData.MotorCADGeo;

    if nargin==3
    ScaledMachineData.refBuildingData=BuildingData;
    end
    k_Axial   =scalingFactorStruct.k_Axial   ;
    k_Radial  =scalingFactorStruct.k_Radial  ;
    k_Winding =scalingFactorStruct.k_Winding ;    

    if isfield(scalingFactorStruct,'turns_per_coil')
    scalingFactorStruct.n_c                 =scalingFactorStruct.turns_per_coil;    
    end
    % ArmatureConductor_Temperature - Emag -% TwindingCalc_MotorLAB 와 동일
    % Twdg_MotorLAB                 - Lab (만들어질때 Emag기준)
    % LabTable=originTable;
%% 형상 K_Winding 정의하는거 추가 (n_c_ref있는지 확인해서)

    l_stk_ref = MotorCADGeo.Stator_Lam_Length; % l_stk,ref의 값
    D_out_ref = MotorCADGeo.Stator_Lam_Dia; % D_out,ref의 값

    if MotorCADGeo.Armature_CoilStyle==0
    n_c_ref  =  MotorCADGeo.MagTurnsConductor; % turn per Coil
    elseif MotorCADGeo.Armature_CoilStyle==1
    n_c_ref  =  MotorCADGeo.WindingLayers; % turn per Coil
    end
    a_p_ref  =  MotorCADGeo.ParallelPaths; % Parallel Path  

    %%
    ScaledMachineData.NumberOfCuboids_LossModel_Lab =MotorCADGeo.NumberOfCuboids_LossModel_Lab;
    ScaledMachineData.NumberStrandsHand =MotorCADGeo.NumberStrandsHand;
    ScaledMachineData.ACConductorLossProportion_Lab=MotorCADGeo.ACConductorLossProportion_Lab;

%  Insulation

    ScaledMachineData.Insulation_Thickness =k_Radial*MotorCADGeo.Insulation_Thickness ;
    ScaledMachineData.Liner_Thickness      =k_Radial*MotorCADGeo.Liner_Thickness      ;
    ScaledMachineData.ConductorSeparation  =k_Radial*MotorCADGeo.ConductorSeparation  ;


%% Stator

    ScaledMachineData.Stator_Lam_Length        = k_Axial * l_stk_ref;
    ScaledMachineData.Stator_Lam_Dia           = k_Radial * D_out_ref;
    ScaledMachineData.Tooth_Tip_Depth          =k_Radial*MotorCADGeo.Tooth_Tip_Depth;
    ScaledMachineData.MinBackIronThickness     =k_Radial*MotorCADGeo.MinBackIronThickness;
%% Axial
% Thermal Housing 

    ScaledMachineData.Rotor_Lam_Length       =ScaledMachineData.Stator_Lam_Length ;
    ScaledMachineData.Stator_Lam_Length      =ScaledMachineData.Stator_Lam_Length ;
    ScaledMachineData.Magnet_Length          =ScaledMachineData.Stator_Lam_Length ;
   
    NonActivePartLength                      = MotorCADGeo.Motor_Length -l_stk_ref;
    ScaledMachineData.Motor_Length           =ScaledMachineData.Stator_Lam_Length+NonActivePartLength;
    NonActivePartDia                         = MotorCADGeo.Housing_Dia  - MotorCADGeo.Stator_Lam_Dia;
    ScaledMachineData.Housing_Dia            =ScaledMachineData.Stator_Lam_Dia   +NonActivePartDia;
%% Detaild Geometric

    MotorCADGeo.MagnetSeparation_Array;
    % ScaledMachineData.MagnetThickness_Array=MachineData.MagnetThickness_Array;
    ScaledMachineData.MagnetSeparation_Array=k_Radial*MotorCADGeo.MagnetSeparation_Array;
    ScaledMachineData.MagnetThickness_Array =k_Radial*MotorCADGeo.MagnetThickness_Array;
%% [TB]Winding
% n_c  turn per Coil
% n_c_per_ap % turn per Coil per ParallelPaths
    ScaledMachineData.n_c_per_ap               = k_Winding * (n_c_ref / a_p_ref);   %  

    if MotorCADGeo.Armature_CoilStyle==0
    ScaledMachineData.MagTurnsConductor        = scalingFactorStruct.n_c;

    elseif MotorCADGeo.Armature_CoilStyle==1
    ScaledMachineData.WindingLayers            = scalingFactorStruct.n_c;
    end
    ScaledMachineData.ParallelPaths            = scalingFactorStruct.a_p;

%% [Geometric]Winding

  if MotorCADGeo.Armature_CoilStyle==0
    % n_c_ref  =  MachineData.MagTurnsConductor; % turn per Coil
  elseif MotorCADGeo.Armature_CoilStyle==1
    ScaledMachineData.Copper_Width =MotorCADGeo.Copper_Width  *k_Radial ;
    ScaledMachineData.Copper_Height=MotorCADGeo.Copper_Height *k_Radial ;
  end
%% 저항은 별도로 End 부분이랑 Active Part 분리 필요
% R_Cuco_ref = SatuMapData.Phase_Resistance_DC_at_20C; % R_Cuco,ref의 값

    R_Cuco_ref              = MotorCADGeo.ResistanceActivePart;
    R_Cuew_ref              = MotorCADGeo.EndWindingResistance_Lab;     % R_Cuew,ref의 값
    % 4 retro
    MotorCADGeo.R_Cuco_ref  = R_Cuco_ref;
    MotorCADGeo.R_Cuew_ref  = R_Cuew_ref;
%% If LabTable

    ScaledMachineData.Twdg_MotorLAB                 = MotorCADGeo.Twdg_MotorLAB;
    ScaledMachineData.ArmatureConductor_Temperature = MotorCADGeo.ArmatureConductor_Temperature;
    ScaledMachineData.TwindingCalc_MotorLAB         = MotorCADGeo.Twdg_MotorLAB;
    % Building Data Needed
    % MotorCADGeo=tempDefMCADMachineData4Scaling(mcad(1))
    % MotorCADGeo.ArmatureConductor_Temperature
    % MotorCADGeo.Twdg_MotorLAB
    % TwindingCalc_MotorLAB
    R_Cuco_ref20=scaleResistancebyTemp(R_Cuco_ref,20,MotorCADGeo.ArmatureConductor_Temperature);
    R_Cuew_ref20=scaleResistancebyTemp(R_Cuew_ref,20,MotorCADGeo.ArmatureConductor_Temperature);
%% 
%% 
% 0.00993


    % 4 retro
    MotorCADGeo.R_Cuco_ref20=R_Cuco_ref20;
    MotorCADGeo.R_Cuew_ref20=R_Cuew_ref20;
    ScaledMachineData.refMachineData=MotorCADGeo;

    ScaledMachineData.scalingFactor =scalingFactorStruct;
    % 20도 기준으로 변경
    ScaledMachineData.ResistanceActivePart20        = (k_Winding^2 / k_Radial^2) * (k_Axial * R_Cuco_ref20);
    ScaledMachineData.EndWindingResistance_Lab20    = (k_Winding^2 / k_Radial^2) * (k_Radial* R_Cuew_ref20);
    ScaledMachineData.Resistance_MotorLAB20         = (k_Winding^2 / k_Radial^2) * (k_Axial * R_Cuco_ref20 + k_Radial * R_Cuew_ref20);
    ScaledMachineData.ArmatureWindingResistancePh20 = (k_Winding^2 / k_Radial^2) * (k_Axial * R_Cuco_ref20 + k_Radial * R_Cuew_ref20);

    %% ArmatureConductor_Temperature 기준으로 변경
    ScaledMachineData.ResistanceActivePart                =scaleResistancebyTemp(ScaledMachineData.ResistanceActivePart20 ,MotorCADGeo.ArmatureConductor_Temperature,20)           ;
    ScaledMachineData.EndWindingResistance_Lab            =scaleResistancebyTemp(ScaledMachineData.EndWindingResistance_Lab20,MotorCADGeo.ArmatureConductor_Temperature,20)        ;
    ScaledMachineData.Resistance_MotorLAB                 =scaleResistancebyTemp(ScaledMachineData.Resistance_MotorLAB20,MotorCADGeo.ArmatureConductor_Temperature,20)               ;
    ScaledMachineData.ArmatureWindingResistancePh         =scaleResistancebyTemp(ScaledMachineData.ArmatureWindingResistancePh20,MotorCADGeo.ArmatureConductor_Temperature,20)  ;     

    %% EndInductance
    ScaledMachineData.EndWindingInductance_Lab            = k_Radial*MotorCADGeo.EndWindingInductance_Lab;

end

File: tools/motorCAD/pyMCAD/scaling/sl_scale_machine.py
"""
sl_scale_machine.py — MATLAB SLScaleMachine 대응

Similarity Law(유사 법칙) 스케일링을 적용하여 ScaledMachineData를 반환한다.

저항 스케일링 법칙:
    R_active(sc) = (kw² / kr²) * ka * R_active(ref, 20°C)
    R_ew(sc)     = (kw² / kr²) * kr * R_ew(ref, 20°C)
"""
from .models import ScalingFactor, MotorCADGeo, ScaledMachineData
from .helpers import scale_resistance_by_temp


def sl_scale_machine(
    factor: ScalingFactor,
    geo: MotorCADGeo,
) -> ScaledMachineData:
    """Similarity Law 스케일링 적용.

    Parameters
    ----------
    factor : ScalingFactor   — def_scaling_factor()로 생성
    geo    : MotorCADGeo     — get_mcad_machine_data()로 생성

    Returns
    -------
    ScaledMachineData

    스케일링 법칙 요약
    ------------------
    형상 (radial)  : × k_radial
    형상 (axial)   : × k_axial
    저항 (active)  : × (k_winding² / k_radial²) × k_axial
    저항 (end-wdg) : × (k_winding² / k_radial²) × k_radial
    인덕턴스 (EW)  : × k_radial
    """
    kr = factor.k_radial
    ka = factor.k_axial
    kw = factor.k_winding

    # ref 권선 파라미터
    n_c_ref = (geo.MagTurnsConductor if geo.Armature_CoilStyle == 0
               else geo.WindingLayers)
    a_p_ref = geo.ParallelPaths

    sd = ScaledMachineData()

    # ── 공통 Lab 속성 (변경 없음) ──
    sd.NumberOfCuboids_LossModel_Lab = geo.NumberOfCuboids_LossModel_Lab
    sd.NumberStrandsHand             = geo.NumberStrandsHand
    sd.ACConductorLossProportion_Lab = geo.ACConductorLossProportion_Lab

    # ── 절연 (radial 스케일) ──
    sd.Insulation_Thickness = kr * geo.Insulation_Thickness
    sd.Liner_Thickness      = kr * geo.Liner_Thickness
    sd.ConductorSeparation  = kr * geo.ConductorSeparation

    # ── 형상 (axial) ──
    l_stk_sc             = ka * geo.Stator_Lam_Length
    sd.Stator_Lam_Length = l_stk_sc
    sd.Rotor_Lam_Length  = l_stk_sc
    sd.Magnet_Length     = l_stk_sc
    sd.Motor_Length      = l_stk_sc + (geo.Motor_Length - geo.Stator_Lam_Length)

    # ── 형상 (radial) ──
    sd.Stator_Lam_Dia       = kr * geo.Stator_Lam_Dia
    sd.Tooth_Tip_Depth      = kr * geo.Tooth_Tip_Depth
    sd.MinBackIronThickness = kr * geo.MinBackIronThickness
    sd.Housing_Dia          = sd.Stator_Lam_Dia + (geo.Housing_Dia - geo.Stator_Lam_Dia)

    # ── 자석 배열 ──
    sd.MagnetSeparation_Array = [kr * v for v in geo.MagnetSeparation_Array]
    sd.MagnetThickness_Array  = [kr * v for v in geo.MagnetThickness_Array]

    # ── 권선 ──
    sd.n_c_per_ap    = kw * (n_c_ref / a_p_ref)
    sd.ParallelPaths = factor.a_p if factor.a_p is not None else a_p_ref

    if geo.Armature_CoilStyle == 0:   # 환선
        sd.MagTurnsConductor = factor.n_c if factor.n_c is not None else n_c_ref
        sd.WindingLayers     = geo.WindingLayers
    else:                              # Hairpin
        sd.WindingLayers = factor.n_c if factor.n_c is not None else n_c_ref
        sd.Copper_Width  = kr * geo.Copper_Width
        sd.Copper_Height = kr * geo.Copper_Height

    # ── 저항 계산 (먼저 20°C 기준으로 환산) ──
    T_op    = geo.ArmatureConductor_Temperature
    R_act20 = scale_resistance_by_temp(geo.ResistanceActivePart,     20.0, T_op)
    R_ew20  = scale_resistance_by_temp(geo.EndWindingResistance_Lab, 20.0, T_op)

    kR = kw ** 2 / kr ** 2
    sd.ResistanceActivePart20        = kR * ka * R_act20
    sd.EndWindingResistance_Lab20    = kR * kr * R_ew20
    sd.Resistance_MotorLAB20         = sd.ResistanceActivePart20 + sd.EndWindingResistance_Lab20
    sd.ArmatureWindingResistancePh20 = sd.Resistance_MotorLAB20

    # ── 동작 온도로 재보정 ──
    sd.ResistanceActivePart      = scale_resistance_by_temp(sd.ResistanceActivePart20,     T_op, 20.0)
    sd.EndWindingResistance_Lab  = scale_resistance_by_temp(sd.EndWindingResistance_Lab20, T_op, 20.0)
    sd.Resistance_MotorLAB       = scale_resistance_by_temp(sd.Resistance_MotorLAB20,      T_op, 20.0)
    sd.ArmatureWindingResistancePh = sd.Resistance_MotorLAB

    # ── 인덕턴스 ──
    sd.EndWindingInductance_Lab = kr * geo.EndWindingInductance_Lab

    # ── 온도 ──
    sd.Twdg_MotorLAB                 = geo.Twdg_MotorLAB
    sd.ArmatureConductor_Temperature = T_op
    sd.TwindingCalc_MotorLAB         = geo.Twdg_MotorLAB

    # ── 참조 데이터 보존 ──
    sd.scalingFactor  = factor
    sd.refMachineData = geo

    return sd
